2010-12-15  H.J. Lu  <hongjiu.lu@intel.com>

	* simple.c (bfd_simple_get_relocated_section_contents): If a
	compressed section hasn't been decompressed, decompress it
	first before applying simple relocations.

diff --git a/bfd/simple.c b/bfd/simple.c
index 03d1a15..7fea092 100644
--- a/bfd/simple.c
+++ b/bfd/simple.c
@@ -207,6 +207,17 @@ bfd_simple_get_relocated_section_contents (bfd *abfd,
       if (data == NULL)
 	return NULL;
       outbuf = data;
+
+      /* If a compressed section hasn't been decompressed, we
+	 decompress it now so that applying simple relocations
+	 won't change its content.  */
+      if (sec->compress_status == DECOMPRESS_SECTION_SIZED)
+	{
+	  contents = NULL;
+	  /* Decompress the section.  */
+	  if (!bfd_get_full_section_contents (abfd, sec, &contents))
+	    return NULL;
+	}
     }
 
   /* The sections in ABFD may already have output sections and offsets set.
